File: annotations.py
#Based on https://github.com/ibaiGorordo/OpenImages-Yolo-converter/blob/master/OIDtoYOLOconverter.py

import csv
import os
import os.path

#Path to the predictions.csv file
predictions_file = '/mnt/c/Users/ccvcauser2/Desktop/open-images-v7/train/labels/detections.csv'
#predictions_file = '/mnt/c/Users/ccvcauser2/Desktop/faceslicenses_dataset_10000/validation/labels/detections.csv'

#Path to the directory containing the images
images_directory = '/mnt/c/Users/ccvcauser2/Desktop/open-images-v7/train/images'
#images_directory = '/mnt/c/Users/ccvcauser2/Desktop/faceslicenses_dataset_10000/validation/data'

labels_directory = '/mnt/c/Users/ccvcauser2/Desktop/open-images-v7/train/labels/'
#labels_directory = '/mnt/c/Users/ccvcauser2/Desktop/faceslicenses_dataset_10000/validation/labels/'

#Path to the directory containing the classes
classes_file = '/mnt/c/Users/ccvcauser2/Desktop/open-images-v7/train/metadata/classes.csv'
#classes_file = '/mnt/c/Users/ccvcauser2/Desktop/faceslicenses_dataset_10000/validation/metadata/classes.csv'

#Path to the output file
#output_file = 'dataset.csv'

""" target_classes = ["Person", "Vehicle registration plate"] """

target_classes = ["Plastic bag", "Waste container", "Billboard", "Traffic sign", "Stop sign"]

""" target_classes = ["Plastic bag", "Waste container", "Billboard", "Traffic sign", "Stop sign",
                        "Sofa bed", "Couch", "Studio couch", "Bed", "Infant bed", "Television",
                        "Refrigerator", "Washing machine", "Home appliance", "Kitchen appliance", 
                        "Shelf", "Cabinetry", "Bathroom cabinet", "Filing cabinet", "Chest of drawers",
                        "Drawer", "Nightstand", "Table", "Coffee table", "Kitchen & dining room table", "Cupboard"] """

#List to store the rows that match the image names
matching_rows = []

target_classes_id = []
annotations_count = [0]*len(target_classes)


def SaveBoundingBoxToFile(image_id,label,x_min,x_max,y_min,y_max):
	
    if os.path.isfile(labels_directory + image_id + ".txt"):
        with open(labels_directory + image_id + ".txt", 'a') as f:
            f.write(' '.join([str(target_classes_id.index(label)),
            #f.write(' '.join([str(1),
                        str(round((x_max+x_min)/2,6)),
                        str(round((y_max+y_min)/2,6)),
                        str(round(x_max-x_min,6)),
                        str(round(y_max-y_min,6))])+'\n')

    else:
            with open(labels_directory + image_id + ".txt", 'w') as f:
                f.write(' '.join([str(target_classes_id.index(label)),
                #f.write(' '.join([str(1),
                            str(round((x_max+x_min)/2,6)),
                            str(round((y_max+y_min)/2,6)),
                            str(round(x_max-x_min,6)),
                            str(round(y_max-y_min,6))])+'\n')        
        


with open(classes_file, 'r') as csvfile:
    reader_classes = csv.reader(csvfile)

    for row_classes in reader_classes:
        
        for class_name in target_classes:

            #print(class_name)
            if row_classes[1] == class_name:  
                target_classes_id.append(row_classes[0])

print("Classes:", target_classes, " ids:", target_classes_id)
            

#Open the predictions.csv file
with open(predictions_file, 'r') as csvfile:
    reader_annotations = csv.reader(csvfile)
    
    # Get the column names from the first row
    #column_names = next(reader_annotations)
    
    # Add the column names to the matching_rows list
    #matching_rows.append(column_names)
    
    # Loop through each row
    for row in reader_annotations:
        
        # Get the image name from the first column
        image_name = row[0] + '.jpg'

        # Check if the image exists in the directory
        if os.path.exists(os.path.join(images_directory, image_name)):
            
            #Check if the annotation class is the one we are looking for
            for class_id in target_classes_id:

                if row[2] == class_id:
                    
                    annotations_count[target_classes_id.index(class_id)] += 1
                    # Add the row to the matching_rows list
                    #matching_rows.append(row)
                    SaveBoundingBoxToFile(row[0], row[2], float(row[4]), float(row[5]), float(row[6]), float(row[7]))
                    
                    break

"""            
                else:

                    if (row[2] == "/m/03bt1vf") or (row[2] == "/m/04yx4") or (row[2] == "/m/0dzct"):    #If it is "Man", "Woman" or "Human face" -> change it to "Person"
   
                        annotations_count[target_classes_id.index(class_id)] += 1
                        # Add the row to the matching_rows list
                        #matching_rows.append(row)
                        SaveBoundingBoxToFile(row[0], "/m/01g317", float(row[4]), float(row[5]), float(row[6]), float(row[7]))
                        break
"""                        

"""
# Write the matching rows to the output file
with open(output_file, 'w', newline='') as csvfile:
    writer = csv.writer(csvfile)
    writer.writerows(matching_rows)
"""

#Split dataset in train, validation and test subsets
from utils.datasets import * 

autosplit(images_directory, [0.7, 0.3, 0.0], annotated_only=False)


print("Finished converting the annotations!")

for i in range(len(target_classes)):
    print("Class: ", target_classes[i], "   N. annotations: ", annotations_count[i])


import os

#os.system("python3 train.py --weights ../Models/yolov7-e6e.pt --data data/faceslicenses.yaml --workers 2 --epochs 50 --batch-size 32 --img 640 --cfg cfg/training/yolov7-e6e_faceslicenses.yaml --project ../Results/yolov7-e6e_5k_tidycity/train")
os.system("python3 train_aux.py --weights ../Models/yolov7-e6.pt --data data/faceslicenses.yaml --workers 1 --epochs 50 --batch-size 8 --img 1280 --cfg cfg/training/yolov7-e6_faceslicenses.yaml --project ../Results/yolov7-e6_5k_peoplelicenses/train")

File: annotations.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

//#define MAX_CLASSES 26
#define MAX_CLASSES 2
#define MAX_CLASS_NAME_LENGTH 50

/* char* target_classes[MAX_CLASSES] = {"Plastic bag", "Waste container", "Billboard", "Traffic sign", "Stop sign",
                        "Sofa bed", "Couch", "Studio couch", "Bed", "Infant bed", "Television",
                        "Refrigerator", "Washing machine", "Home appliance", "Kitchen appliance", 
                        "Shelf", "Cabinetry", "Bathroom cabinet", "Filing cabinet", "Chest of drawers",
                        "Drawer", "Nightstand", "Table", "Coffee table", "Kitchen & dining room table", "Cupboard"};
 */
char* target_classes[MAX_CLASSES] ={"Person", "Vehicle registration plate"};

int target_classes_id[MAX_CLASSES];
int annotations_count[MAX_CLASSES] = {0};

char* images_directory = "/mnt/c/Users/ccvcauser2/Desktop/open-images-v7/train/data";
char* labels_directory = "/mnt/c/Users/ccvcauser2/Desktop/open-images-v7/train/labels/";
char* classes_file = "/mnt/c/Users/ccvcauser2/Desktop/open-images-v7/train/metadata/classes.csv";
char* predictions_file = "/mnt/c/Users/ccvcauser2/Desktop/open-images-v7/train/labels/detections.csv";


void SaveBoundingBoxToFile(char* image_id, int label, float x_min, float x_max, float y_min, float y_max) {
    char file_path[200];
    sprintf(file_path, "%s%s.txt", labels_directory, image_id);

    FILE* f;
    if ((f = fopen(file_path, "a")) != NULL) {
        fprintf(f, "%d %f %f %f %f\n", label, (x_max + x_min) / 2, (y_max + y_min) / 2, x_max - x_min, y_max - y_min);
        fclose(f);
    }
    else if ((f = fopen(file_path, "w")) != NULL) {
        fprintf(f, "%d %f %f %f %f\n", label, (x_max + x_min) / 2, (y_max + y_min) / 2, x_max - x_min, y_max - y_min);
        fclose(f);
    }
    else {
        printf("Failed to open file %s\n", file_path);
    }
}

void ReadClassesFile() {
    FILE* f;
    if ((f = fopen(classes_file, "r")) != NULL) {
        char line[200];
        char* ptr;
        while (fgets(line, sizeof(line), f) != NULL) {
            ptr = strtok(line, ",");
            int id = atoi(ptr);
            ptr = strtok(NULL, ",");
            ptr[strcspn(ptr, "\r\n")] = 0;
            for (int i = 0; i < MAX_CLASSES; i++) {
                if (strcmp(ptr, target_classes[i]) == 0) {
                    target_classes_id[i] = id;
                    break;
                }
            }
        }
        fclose(f);
    }
    else {
        printf("Failed to open file %s\n", classes_file);
    }
}

void ReadPredictionsFile() {
    FILE* f;
    if ((f = fopen(predictions_file, "r")) != NULL) {
        char line[200];
        char* ptr;
        int row_count = 0;
        while (fgets(line, sizeof(line), f) != NULL) {
            if (row_count != 0) {
                
                ptr = strtok(line, ",");
                char image_name[100];
                strcpy(image_name, ptr);

                ptr = strtok(NULL, ",");
                int class_id = atoi(ptr);

                ptr = strtok(NULL, ",");
                float confidence = atof(ptr);

                ptr = strtok(NULL, ",");
                float x_min = atof(ptr);

                ptr = strtok(NULL, ",");
                float x_max = atof(ptr);

                ptr = strtok(NULL, ",");
                float y_min = atof(ptr);

                ptr = strtok(NULL, ",");
                float y_max = atof(ptr);

                for (int i = 0; i < MAX_CLASSES; i++) {
                    if (class_id == target_classes_id[i]) {
                        SaveBoundingBoxToFile(image_name, i, x_min, x_max, y_min, y_max);
                        annotations_count[i]++;
                        break;
                    }
                }           
                
            }
        }
    }
}

int main() {
    ReadClassesFile();
    printf("Classes: %s %s", target_classes[0], target_classes[1]);
    printf("Ids: %d %d", target_classes_id[0], target_classes_id[1]);

    ReadPredictionsFile();

    for (int i = 0; i < MAX_CLASSES; i++) {
        printf("Class: %s, Count: %d\n", target_classes[i], annotations_count[i]);
    }

    return 0;
}